### PR TITLE
[Vi-Mode] Supports di' and di" commands (quoted text objects)

### LINKED ISSUE #3790: [vi-mode] Support di' and di" commands
### Prerequisites

- [X] Write a descriptive title.

### Description of the new feature/enhancement

As alluded to in #584, PSRL should support vim "text-objects".
Consider supporting the <kbd>d</kbd><kbd>i</kbd><kbd>'</kbd> and <kbd>d</kbd><kbd>i</kbd><kbd>"</kbd> commands to delete quoted text.

### Proposed technical implementation details (optional)

Most of the ~~[ground work](https://github.com/PowerShell/PSReadLine/pull/2059)~~ [ground work](https://github.com/PowerShell/PSReadLine/pull/3853) done to support Text Objects with respect to key chords has already been laid. This feature will be very easy to implement and only involves wiring a couple of new keybindings.

### PR DESCRIPTION
**Pull request by springcomp**
# PR Summary

Fixes #3790.

## PR Checklist

- [x] PR has a meaningful title
    - Use the present tense and imperative mood when describing your changes
- [x] Summarized changes
- [x] Make sure you've added one or more new tests
- [x] Make sure you've tested these changes in terminals that PowerShell is commonly used in (i.e. conhost.exe, Windows Terminal, Visual Studio Code Integrated Terminal, etc.)
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] Documentation needed at [PowerShell-Docs](https://github.com/MicrosoftDocs/PowerShell-Docs)
        - [ ] Doc Issue filed: 

As #2059 has been readily accepted, we have not had the time to discuss the documentation.
This PR, like #2059, wire keybindings to private methods. Please, let me know how to best document this behaviour.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/PowerShell/PSReadLine/pull/3791)

### CODE REVIEW

File: PSReadLine/StringBuilderTextObjectExtensions.cs [hunk 1]
```diff
@@ -109,5 +110,161 @@ public static int ViFindBeginningOfNextWordObjectBoundary(this StringBuilder buf
...
+
+            return (start, end);
+        }
+
+        /// <summary>
+        /// Returns the position of the beginning of line
+        /// starting from the specified "current" position.
+        /// </summary>
+        /// <param name="current">The position in the current logical line.</param>
+        internal static int GetBeginningOfLogicalLinePos(this StringBuilder buffer, int current)
```
Comment (daxian-dbw) [inline PSReadLine/StringBuilderTextObjectExtensions.cs:232]: It seems `GetBeginningOfLogicalLinePos` and `GetEndOfLogicalLinePos` shoudl be put in `StringBuilderLinewiseExtensions` instead.

File: PSReadLine/StringBuilderTextObjectExtensions.cs [hunk 2]
```diff
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Management.Automation;
```
Comment (daxian-dbw) [inline PSReadLine/StringBuilderTextObjectExtensions.cs:2]: Added by mistake? It doesn't seem to be needed by the code changes in this file.

**Comment by daxian-dbw** [comment]
Before moving forward with more 3-key chord supports, we need to first figure out how to make `Get/Set-PSReadLineKeyHandler` support chords with more than 2 keys. Today, `_chordDispatchTable` is a 2-level dictionary, and hence it only really supports 2-key chords. `Get/Set-PSReadLineKeyHandler` are implemented under that assumption.

By looking at #2059 again, I think I merged it a little too prematurely :)
Today, we have two 3-key chords built-in:
- `<d,g,g>` -- implemented by having `g` mapped to `ViDBChord` in the `DTable` with a `DGTable` that maps the second `g` to the handler.
- `<d,w,i>` --implemented by #2059.

There are 2 concerns here:
1. Their implementation should be consistent in the way how the 3rd level dictionary is stored.
2. Changes to VI key binding data structure are needed to make `Get/Set-PSReadLineKeyHandler` to support chords with 3 keys (are we stopping at 3-key chords for VI mode here? maybe we should and not support chords with more than 3 keys).

I think we should first resolve the above concerns before adding more text object commands.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
Before moving forward with more 3-key chord supports, we need to first figure out how to make `Get/Set-PSReadLineKeyHandler` support chords with more than 2 keys. Today, `_chordDispatchTable` is a 2-level dictionary, and hence it only really supports 2-key chords. `Get/Set-PSReadLineKeyHandler` are implemented under that assumption.

By looking at #2059 again, I think I merged it a little too prematurely :)
Today, we have two 3-key chords built-in:
- `<d,g,g>` -- implemented by having `g` mapped to `ViDBChord` in the `DTable` with a `DGTable` that maps the second `g` to the handler.
- `<d,w,i>` --implemented by #2059.

There are 2 concerns here:
1. Their implementation should be consistent in the way how the 3rd level dictionary is stored.
2. Changes to VI key binding data structure are needed to make `Get/Set-PSReadLineKeyHandler` to support chords with 3 keys (are we stopping at 3-key chords for VI mode here? maybe we should and not support chords with more than 3 keys).

I think we should first resolve the above concerns before adding more text object commands.

**Comment by daxian-dbw** [comment]
BTW, the `<d,i>` shows up in the results of `Get-PSReadLineKeyHandler`:
```
User defined functions
======================

Key   Function          Description
---   --------          -----------
<d,i> ChordViTextObject ChordViTextObject
```

Two problems here:
1. `ChordViTextObject` is not put in `GetDisplayGrouping(string function)`, so it's displayed as **User defined functions**.
2. The chord `<d,i>` is incomplete -- users will try using it but observe no result (or unexpected result).

It's the same for `<d,g>`:
```
Miscellaneous functions
=======================

Key        Function          Description
---        --------          -----------
<d,g>      ViDGChord         ViDGChord
```

**Comment by springcomp** [comment]
@daxian-dbw  I have published #3853 to address your feedback, re-working #2059 in the process.
I’m happy for this current PR to be closed as I will be able to move the code forward in a better shape if #3853 is accepted.

**Pull request closed without merging**